{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyNwLiqstPMAlqKjTFwx4i7o"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"acXaJ_dlbLr1","executionInfo":{"status":"ok","timestamp":1743883715074,"user_tz":420,"elapsed":786,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"7506bc77-18e3-4525-afab-358e4dc3dc9d"},"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}],"source":["# connect to google drive to access the data files\n","# You will need to give premission to access your drive\n","# if this is the first time doing so\n","\n","from google.colab import drive\n","drive.mount('/content/drive')\n","\n","# Create paths to file locations\n","# Update these paths to where your store the files in your drive\n","Lead_path = '/content/drive/MyDrive/LSC 586/Arabidopsis Lead Treatment.gz'\n","Control_Path = '/content/drive/MyDrive/LSC 586/Arabidopsis Untreated Control.gz'\n","Vanadium_Path = '/content/drive/MyDrive/LSC 586/Arabidopsis Vanadium Treatment.gz'\n","\n","fasta_path = '/content/drive/MyDrive/LSC 586/GCF_000001735.4_TAIR10.1_genomic.fna'\n","gtf_gz_path = '/content/drive/MyDrive/LSC 586/GCF_000001735.4_TAIR10.1_genomic.gtf.gz'"]},{"cell_type":"markdown","source":["## Installtion with STAR\n","\n","STAR is an ultra fast universal RNA-seq alignment tool\n","\n","Paper: https://pmc.ncbi.nlm.nih.gov/articles/PMC3530905/\n","\n","### Tutorial:\n","\n","Here: https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/03_alignment.html"],"metadata":{"id":"MdkBKp_fIT9F"}},{"cell_type":"code","source":["# install STAR\n","\n","# Step 1: Download the STAR source tarball\n","star_tarball_url = \"https://github.com/alexdobin/STAR/archive/2.7.9a.tar.gz\"\n","tarball_name = \"STAR-2.7.9a.tar.gz\"\n","\n","print(\"Downloading STAR source code...\")\n","subprocess.run([\"wget\", star_tarball_url, \"-O\", tarball_name], check=True)\n","\n","# Step 2: Extract the tarball\n","print(\"Extracting STAR source code...\")\n","subprocess.run([\"tar\", \"-xzvf\", tarball_name], check=True)\n","\n","# Step 3: Compile STAR\n","star_source_dir = \"STAR-2.7.9a/source\"\n","print(\"Compiling STAR...\")\n","subprocess.run([\"make\", \"-C\", star_source_dir], check=True)\n","\n","# Confirm that the STAR binary exists\n","star_binary = os.path.join(star_source_dir, \"STAR\")\n","if os.path.exists(star_binary):\n","    print(f\"STAR compiled successfully! Binary is located at: {star_binary}\")\n","else:\n","    print(\"Compilation failed: STAR binary not found.\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2x2x5akekuTS","executionInfo":{"status":"ok","timestamp":1743884428667,"user_tz":420,"elapsed":321938,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"78c606e0-ab90-4408-9716-be7dea542cd5"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Downloading STAR source code...\n","Extracting STAR source code...\n","Compiling STAR...\n","STAR compiled successfully! Binary is located at: STAR-2.7.9a/source/STAR\n"]}]},{"cell_type":"code","source":["!apt-get install -y samtools subread"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"t0TNIJilHGzk","executionInfo":{"status":"ok","timestamp":1743894367181,"user_tz":420,"elapsed":4644,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"49e378f8-7ace-423d-92b4-fdf1a974d1b6"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Reading package lists... Done\n","Building dependency tree... Done\n","Reading state information... Done\n","subread is already the newest version (2.0.3+dfsg-1).\n","0 upgraded, 0 newly installed, 0 to remove and 44 not upgraded.\n"]}]},{"cell_type":"code","source":["# import required packages\n","import gzip\n","import shutil\n","import os\n","import subprocess"],"metadata":{"id":"VJ5Z6nZLe8l7"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# 1. Unzip and decompress Data sets"],"metadata":{"id":"aW0uRh67I2Ct"}},{"cell_type":"code","source":["def decompress_gz(input_path, output_path):\n","    \"\"\"Decompress a .gz file to the specified output file.\"\"\"\n","    with gzip.open(input_path, 'rb') as f_in:\n","        with open(output_path, 'wb') as f_out:\n","            shutil.copyfileobj(f_in, f_out)\n","    print(f\"Decompressed {input_path} to {output_path}\")\n","\n","# Create a directory for decompressed data\n","os.makedirs(\"raw_data\", exist_ok=True)\n","\n","# Define gzipped files and desired output FASTQ\n","files = {\n","    Lead_path: \"Arabidopsis_Lead_Treatment.fastq\",\n","    Control_Path: \"Arabidopsis_Untreated_Control.fastq\",\n","    Vanadium_Path: \"Arabidopsis_Vanadium_Treatment.fastq\"\n","}\n","\n","# Decompress and save to raw_data directory\n","for gz_file, fastq_file in files.items():\n","    input_path = gz_file\n","    output_path = os.path.join(\"raw_data\", fastq_file)\n","    decompress_gz(input_path, output_path)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fordX8APfzqM","executionInfo":{"status":"ok","timestamp":1743882784700,"user_tz":420,"elapsed":182319,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"497369a0-529e-4640-bc57-c60e886784ca"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Decompressed /content/drive/MyDrive/LSC 586/Arabidopsis Lead Treatment.gz to raw_data/Arabidopsis_Lead_Treatment.fastq\n","Decompressed /content/drive/MyDrive/LSC 586/Arabidopsis Untreated Control.gz to raw_data/Arabidopsis_Untreated_Control.fastq\n","Decompressed /content/drive/MyDrive/LSC 586/Arabidopsis Vanadium Treatment.gz to raw_data/Arabidopsis_Vanadium_Treatment.fastq\n"]}]},{"cell_type":"markdown","source":["# 2. Create Index for alignment on Reference Genome"],"metadata":{"id":"-xkTiffcJT6y"}},{"cell_type":"code","source":["def decompress_gtf(gz_file, output_file):\n","    \"\"\"\n","    Decompress a gzipped GTF file.\n","    \"\"\"\n","    with gzip.open(gz_file, 'rb') as f_in:\n","        with open(output_file, 'wb') as f_out:\n","            shutil.copyfileobj(f_in, f_out)\n","    print(f\"Decompressed {gz_file} to {output_file}\")\n","\n","# Define file paths\n","fasta_file = fasta_path\n","gtf_gz_file = gtf_gz_path\n","gtf_file = \"GCF_000001735.4_TAIR10.1_genomic.gtf\"  # Decompressed file output\n","star_index_dir = \"star_index\"\n","\n","# Check if the GTF file has been decompressed\n","if not os.path.exists(gtf_file):\n","    decompress_gtf(gtf_gz_file, gtf_file)\n","\n","# Create the STAR index directory if it doesn't exist\n","# This step is crucial for STAR to work properly\n","os.makedirs(star_index_dir, exist_ok=True)\n","\n","# Define parameters\n","threads = \"4\"\n","sjdbOverhang = \"99\"  # read length minus one but we can adjust this as needed\n","\n","# Build the STAR index command\n","star_command = [\n","    \"STAR-2.7.9a/source/STAR\",\n","    \"--runThreadN\", threads,\n","    \"--runMode\", \"genomeGenerate\",\n","    \"--genomeDir\", star_index_dir,\n","    \"--genomeFastaFiles\", fasta_file,\n","    \"--sjdbGTFfile\", gtf_file,\n","    \"--sjdbOverhang\", sjdbOverhang\n","]\n","\n","print(\"Building STAR index with the following command:\")\n","print(\" \".join(star_command))\n","\n","# Run STAR command to generate the index\n","subprocess.run(star_command, check=True)\n","\n","print(\"STAR index built successfully!\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"y8iNr1ZckEzu","executionInfo":{"status":"ok","timestamp":1743885143782,"user_tz":420,"elapsed":666687,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"5659a89d-3ba0-4ff6-ae7c-c7c5350da249"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Building STAR index with the following command:\n","STAR-2.7.9a/source/STAR --runThreadN 4 --runMode genomeGenerate --genomeDir star_index --genomeFastaFiles /content/drive/MyDrive/LSC 586/GCF_000001735.4_TAIR10.1_genomic.fna --sjdbGTFfile GCF_000001735.4_TAIR10.1_genomic.gtf --sjdbOverhang 99\n","STAR index built successfully!\n"]}]},{"cell_type":"markdown","source":["# 3. Align data samples to indexed reference genome"],"metadata":{"id":"LGsJxb6lJb1k"}},{"cell_type":"code","source":["# This cell will take approximately 2 hours to run given the 3 files below\n","\n","# List of FASTQ files (from \"raw_data\" directory)\n","samples = [\n","    \"Arabidopsis_Lead_Treatment.fastq\",\n","    \"Arabidopsis_Untreated_Control.fastq\",\n","    \"Arabidopsis_Vanadium_Treatment.fastq\"\n","]\n","\n","raw_data_dir = \"raw_data\"\n","alignment_dir = \"alignments_star\"\n","os.makedirs(alignment_dir, exist_ok=True)\n","threads = \"4\"\n","star_index_dir = \"star_index\"\n","\n","# The absolute path to your STAR binary (from the previous step)\n","star_binary = \"./STAR-2.7.9a/source/STAR\"\n","\n","for sample in samples:\n","    fastq_path = os.path.join(raw_data_dir, sample)\n","    # Set an output prefix for STAR\n","    output_prefix = os.path.join(alignment_dir, sample.replace(\".fastq\", \"_\"))\n","\n","    # Build STAR command for aligning the reads\n","    star_command = [\n","        star_binary,\n","        \"--runThreadN\", threads,\n","        \"--genomeDir\", star_index_dir,\n","        \"--readFilesIn\", fastq_path,\n","        \"--outFileNamePrefix\", output_prefix\n","    ]\n","\n","    print(f\"Aligning sample {sample}...\")\n","    subprocess.run(star_command, check=True)\n","    print(f\"Alignment for {sample} completed. Output files begin with: {output_prefix}\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pzgzXQaSsTVR","executionInfo":{"status":"ok","timestamp":1743892617600,"user_tz":420,"elapsed":6804773,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"77fbb0b2-006a-49eb-b55a-ab5106fe212c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Aligning sample Arabidopsis_Lead_Treatment.fastq...\n","Alignment for Arabidopsis_Lead_Treatment.fastq completed. Output files begin with: alignments_star/Arabidopsis_Lead_Treatment_\n","Aligning sample Arabidopsis_Untreated_Control.fastq...\n","Alignment for Arabidopsis_Untreated_Control.fastq completed. Output files begin with: alignments_star/Arabidopsis_Untreated_Control_\n","Aligning sample Arabidopsis_Vanadium_Treatment.fastq...\n","Alignment for Arabidopsis_Vanadium_Treatment.fastq completed. Output files begin with: alignments_star/Arabidopsis_Vanadium_Treatment_\n"]}]},{"cell_type":"markdown","source":["# 4. Convert SAM to BAM\n","\n","After the alignment, SAM files will be created. We need to convert these to BAM files and sort them in order to get a gene expression count"],"metadata":{"id":"DCwLzOZtJ3De"}},{"cell_type":"code","source":["# Directories for alignment outputs and BAM files\n","alignment_dir = \"alignments_star\"\n","bam_dir = \"bam_files\"\n","os.makedirs(bam_dir, exist_ok=True)\n","\n","# List SAM file names as generated by STAR\n","# These should be located in the alignments_star folder that was created.\n","# If not update below to their paths\n","samples = [\n","    \"/content/alignments_star/Arabidopsis_Lead_Treatment_Aligned.out.sam\",\n","    \"/content/alignments_star/Arabidopsis_Untreated_Control_Aligned.out.sam\",\n","    \"/content/alignments_star/Arabidopsis_Vanadium_Treatment_Aligned.out.sam\"\n","]\n","\n","# Convert SAM to BAM and sort\n","for sample in samples:\n","    sam_file = os.path.join(alignment_dir, sample)\n","    bam_file = os.path.join(bam_dir, sample.replace(\".sam\", \".bam\"))\n","    sorted_bam_file = os.path.join(bam_dir, sample.replace(\".sam\", \"_sorted.bam\"))\n","\n","    # Convert SAM to BAM\n","    print(f\"Converting {sam_file} to BAM...\")\n","    subprocess.run([\"samtools\", \"view\", \"-bS\", sam_file, \"-o\", bam_file], check=True)\n","\n","    # Sort the BAM file\n","    print(f\"Sorting {bam_file}...\")\n","    subprocess.run([\"samtools\", \"sort\", bam_file, \"-o\", sorted_bam_file], check=True)\n","\n","    print(f\"Finished processing {sample}. Sorted BAM: {sorted_bam_file}\")\n","\n","# Count reads per gene using featureCounts\n","# List all sorted BAM files\n","sorted_bam_files = [os.path.join(bam_dir, sample.replace(\".sam\", \"_sorted.bam\")) for sample in samples]\n","\n","# annotation file (GTF) and output file for gene counts\n","annotation_file = \"GCF_000001735.4_TAIR10.1_genomic.gtf\"\n","output_counts = \"gene_counts.txt\"\n","\n","# Build the featureCounts command:\n","# -T: number of threads\n","# -a: annotation file (GTF)\n","# -o: output file for counts\n","# Followed by a list of BAM files.\n","featurecounts_command = [\n","    \"featureCounts\",\n","    \"-T\", \"4\",\n","    \"-a\", annotation_file,\n","    \"-o\", output_counts\n","] + sorted_bam_files\n","\n","print(\"Running featureCounts to generate gene count matrix...\")\n","subprocess.run(featurecounts_command, check=True)\n","print(f\"Gene counts generated and saved in {output_counts}\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mM_1AIXOG29b","executionInfo":{"status":"ok","timestamp":1743894174493,"user_tz":420,"elapsed":1144007,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"a092d727-3d5a-4082-ac58-8d738ca956a8"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Converting /content/alignments_star/Arabidopsis_Lead_Treatment_Aligned.out.sam to BAM...\n","Sorting /content/alignments_star/Arabidopsis_Lead_Treatment_Aligned.out.bam...\n","Finished processing /content/alignments_star/Arabidopsis_Lead_Treatment_Aligned.out.sam. Sorted BAM: /content/alignments_star/Arabidopsis_Lead_Treatment_Aligned.out_sorted.bam\n","Converting /content/alignments_star/Arabidopsis_Untreated_Control_Aligned.out.sam to BAM...\n","Sorting /content/alignments_star/Arabidopsis_Untreated_Control_Aligned.out.bam...\n","Finished processing /content/alignments_star/Arabidopsis_Untreated_Control_Aligned.out.sam. Sorted BAM: /content/alignments_star/Arabidopsis_Untreated_Control_Aligned.out_sorted.bam\n","Converting /content/alignments_star/Arabidopsis_Vanadium_Treatment_Aligned.out.sam to BAM...\n","Sorting /content/alignments_star/Arabidopsis_Vanadium_Treatment_Aligned.out.bam...\n","Finished processing /content/alignments_star/Arabidopsis_Vanadium_Treatment_Aligned.out.sam. Sorted BAM: /content/alignments_star/Arabidopsis_Vanadium_Treatment_Aligned.out_sorted.bam\n","Running featureCounts to generate gene count matrix...\n","Gene counts generated and saved in gene_counts.txt\n"]}]},{"cell_type":"markdown","source":["# 5. Generate gene count file to be used in Differential Expression Analysis"],"metadata":{"id":"xXiGppGOKqIK"}},{"cell_type":"code","source":["# Directory where sorted BAM files are stored\n","bam_dir = \"/content/alignments_star\"\n","\n","# List sorted BAM files (assumes they end with \"_sorted.bam\")\n","sorted_bam_files = [os.path.join(bam_dir, f) for f in os.listdir(bam_dir) if f.endswith(\"_sorted.bam\")]\n","print(\"Sorted BAM files:\", sorted_bam_files)\n","\n","# annotation file (GTF format)\n","annotation_file = \"GCF_000001735.4_TAIR10.1_genomic.gtf\"\n","\n","# output file for gene counts\n","output_counts = \"gene_counts.txt\"\n","\n","# Build the featureCounts command:\n","# -T: Number of threads to use\n","# -a: Annotation file (GTF)\n","# -o: Output file for the count matrix\n","featurecounts_command = [\n","    \"featureCounts\",\n","    \"-T\", \"4\",\n","    \"-a\", annotation_file,\n","    \"-o\", output_counts\n","] + sorted_bam_files\n","\n","print(\"Running featureCounts with the following command:\")\n","print(\" \".join(featurecounts_command))\n","\n","# Generate gene counts\n","try:\n","    subprocess.run(featurecounts_command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n","    print(f\"Gene counts generated successfully and saved in {output_counts}\")\n","except subprocess.CalledProcessError as e:\n","    print(\"Error during featureCounts execution:\")\n","    print(e.stderr.decode())\n","\n","# OPTIONAL: Save results to csv file\n","# Read the gene counts file while skipping comment lines (those starting with '#')\n","df = pd.read_csv(output_counts, sep=\"\\t\", comment=\"#\")\n","\n","# Write the DataFrame to a CSV file\n","df.to_csv(\"gene_counts.csv\", index=False)\n","print(\"Gene counts saved to gene_counts.csv\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"OuF1cJihHbjR","executionInfo":{"status":"ok","timestamp":1743894603487,"user_tz":420,"elapsed":115895,"user":{"displayName":"Harrison Stuckey","userId":"08438511478587303872"}},"outputId":"b7411d32-2a18-482a-d0ae-784b8fe2e516"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Sorted BAM files: ['/content/alignments_star/Arabidopsis_Untreated_Control_Aligned.out_sorted.bam', '/content/alignments_star/Arabidopsis_Lead_Treatment_Aligned.out_sorted.bam', '/content/alignments_star/Arabidopsis_Vanadium_Treatment_Aligned.out_sorted.bam']\n","Running featureCounts with the following command:\n","featureCounts -T 4 -a GCF_000001735.4_TAIR10.1_genomic.gtf -o gene_counts.txt /content/alignments_star/Arabidopsis_Untreated_Control_Aligned.out_sorted.bam /content/alignments_star/Arabidopsis_Lead_Treatment_Aligned.out_sorted.bam /content/alignments_star/Arabidopsis_Vanadium_Treatment_Aligned.out_sorted.bam\n","Gene counts generated successfully and saved in gene_counts.txt\n"]}]}]}